lib/kafka_ex/server_0_p_9_p_0.ex

```diff
@@ -73,96 +73,92 @@ defmodule KafkaEx.Server0P9P0 do
     {:ok, state}
   end
 
-  def kafka_server_join_group(join_group_request, network_timeout, state) do
-    true = consumer_group?(state)
-    {broker, state} = broker_for_consumer_group_with_update(state)
-    wire_request = JoinGroup.create_request(
-      state.correlation_id,
-      @client_id,
-      join_group_request
+  def kafka_server_join_group(request, network_timeout, state_in) do
+    true = consumer_group?(state_in)
```

Comment:
it seems like we could drop this `true = consumer_group?(state_in)` in each of these functions and just pattern match in `consumer_group_sync_request/4`? This would also raise a match error, rather than giving a decent failure, which we could do in the shared function.
e.g.
```
def consumer_group_sync_request(_req, _mod, _timeout, %State{consumer_group: :no_consumer_group}) do
  raise "You are trying to consumer group but you aren't consumer grouping"
end
```

Reply:
Good idea.  At first I had this in the shared function, but I wanted it to error out in the parent so that you would see that in the stacktrace.  Making it raise an error and decorating the error with info from the request struct is even better :+1: